### PR TITLE
#164694219 Enable social login via google and facebook

### PR DESCRIPTION
**Pull request by richien**
### What does this PR do?

Enable social users to log into the application using Google, and Facebook

### Description of Task to be completed?

* Ensure registered users can log in using their Google, and Facebook accounts.
* Enable even unregistered users to register and log in with their Google, and Facebook
accounts.
Have the following endpoints working.
* POST `/api/social-auth/google`
* POST `/api/social-auth/facebook`

### How should this be manually tested?

* After cloning this repo, cd into it and switch to this branch
* Install the requirements and set the environment variables as in the `.env.example` file
* Start the development server using `python manage.py runserver`

#### Testing Facebook Social Authentication.
1. Create an application on [facebook developer page](https://developers.facebook.com/).
2. Then get the access_token [here](https://developers.facebook.com/tools/explorer).
3. Test using postman.
POST /api/social-auth/facebook

Request body
```
{
        "access_token": "your_facebook_access_token"
}
```
Response
```
{
    "user": {
        "email": "wilgob2001@yahoo.com",
        "username": "Geoffrey Willis Barugahare",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNuIjoicmVpZnJlZCIsImV4cCI6MTU1NzgzNjA4OH0.Q0hgBaJGTAqINl1gPe9BzFO4xtwtm_3QW9yWk3Zw4c8"
    }
}
```
#### Testing Google Social Authentication.
1. Obtain google user_id [here](https://developers.google.com/oauthplayground/).
2. In the scope selection, search for Google OAuth2 API v2 and check all of its three fields.
googlepage
3. Then click the Authorize APIs button
4. Once you have authorized Google+ API to access your information, click Exchange authorization for codes which will give you an access_token and id_token. Copy the id_token and paste it in Postman endpoint body for google login.
token
5. Test using postman.
POST /api/social-auth/google

Request body
```
{
        "access_token": "your_google_user_id_token"
}
```
Response
```
{
    "user": {
        "email": "benwill2014@gmail.com",
        "username": "ben will",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MSwidXNuIjoicmVpZnJlZCIsImV4cCI6MTU1NzgzNjA4OH0.Q0hgBaJGTAqINl1gPe9BzFO4xtwtm_3QW9yWk3Zw4c8"
    }
}
```

#### Any background context you want to provide?

- This Implementation supports only Google and Facebook authentications and not Twitter.

## What are the relevant pivotal tracker stories?
[#164691219](https://www.pivotaltracker.com/story/show/164691219)

### Screenshots
image
image

### CODE REVIEW

**Review by NicholusMuwonge — CHANGES_REQUESTED** [review]
Hey @richien please squash those commits into one as I look into the rest of the implementation.

**Comment by richien** [comment]
> Hey @richien please squash those commits into one as I look into the rest of the implementation.

Hi @NicholusMuwonge, thank you for the feedback. This has been done.

**Review by frankopkusianwar — APPROVED** [review]
LGTM

**Review by NicholusMuwonge — APPROVED** [review]
Works fine, as projected.

**Review by sebbss — APPROVED** [review]

File: authors/apps/social_auth/login_register.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            "username": user.username,
+            "token": user.auth_token()
+        }
+    except User.DoesNotExist:
+        if social_user.get("email") is None:
+            raise ValidationError("Users must have an email address.")
+
+        new_social_user = {
+            "email": social_user.get("email"),
+            "username": social_user.get("name", "unknown"),
```
Comment (kimbugp) [inline authors/apps/social_auth/login_register.py]: @richien, what would happen if more than one user signing up has no username field yet it is a unique field.
I suggest you should find a way to create a unique username which they can edit later

File: authors/apps/social_auth/login_register.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            "password": User.objects.make_random_password()
+        }
+
+        new_user = User.objects.create_user(**new_social_user)
+        new_user.is_active = True
+        new_user.save()
+
+        return {
```
Comment (kimbugp) [inline authors/apps/social_auth/login_register.py]: This seems to be repeated with code on line 9.
you can find a way to write it one

File: authors/apps/social_auth/login_register.py (hunk 1 above)
Comment (richien) [reply authors/apps/social_auth/login_register.py]: Thank you for the feedback. This has been implemented.

File: authors/apps/social_auth/login_register.py (hunk 2 above)
Comment (richien) [reply authors/apps/social_auth/login_register.py]: Thank you for the feedback. This has been implemented.

**Pull request merged**